orttraining/orttraining/core/graph/pipeline_transformer.cc

```diff
@@ -0,0 +1,255 @@
+  // backward node's input.
+  auto add_leaf_node = [&](std::vector<NodeArg*>& input_args) {
+    for (auto& node : graph.Nodes()) {
+      if (!IsBackward(node)) {
+        // only check backward node
+        continue;
+      }
+      bool find_consumer_nodes = false;
+      std::vector<NodeArg*>& outputs = node.MutableOutputDefs();
+      for (auto& output : outputs) {
+        std::vector<const Node*> consumer_nodes = graph.GetConsumerNodes(output->Name());
+        if (consumer_nodes.size() > 0) {
+          find_consumer_nodes = true;
+          break;
+        }
+      }
+      if (!find_consumer_nodes && outputs.size() > 0) {
+        input_args.push_back(outputs[0]);
+      }
+    }
+  };
+
+  std::vector<NodeArg*> input_args;
+  AddInputEvent(graph, "RecordEvent", false /* is_forward */, input_args, new_input_names);
+
+  if (send_bw) {
+    // if we have send op in backward pass (at the end of the graph), add the RecordEvent op after that.
+    input_args.insert(std::end(input_args),
+                      std::begin(send_bw->MutableOutputDefs()),
+                      std::end(send_bw->MutableOutputDefs()));
+  }
+  add_leaf_node(input_args);
+  graph.AddNode(graph.GenerateNodeName("RecordEvent"),
+                "RecordEvent",
+                "Backward pass",
+                input_args,
+                {},
+                nullptr,
+                kMSDomain);
+}
+
+Status AddWaitForward(Graph& graph, Node* recv_fw, std::vector<std::string>& new_input_names) {
+  auto update_wait_input_output = [&](NodeArg* old_input,
+                                      std::vector<NodeArg*>& input_args,
+                                      std::vector<NodeArg*>& output_args) -> NodeArg& {
+    const auto& new_name = graph.GenerateNodeArgName(old_input->Name());
+    ONNX_NAMESPACE::TypeProto type_proto(*(old_input->TypeAsProto()));
+
+    auto& wait_output = graph.GetOrCreateNodeArg(new_name, &type_proto);
+    output_args.push_back(&wait_output);
+    input_args.push_back(old_input);
+    return wait_output;
+  };
+
+  if (recv_fw) {
+    // if we have recv op in forward pass (at the begining of the graph), add the WaitEvent op before that.
+    std::vector<NodeArg*> input_args;
+    std::vector<NodeArg*> output_args;
+    AddInputEvent(graph, "WaitEvent", true /* is_forward */, input_args, new_input_names);
+
+    // recv's first input is the signal input. Re-direct it to WaitEvent's input.
+    auto& input_signal = recv_fw->MutableInputDefs()[0];
+    auto& wait_output = update_wait_input_output(input_signal, input_args, output_args);
+    input_signal = &wait_output;
+
+    graph.AddNode(graph.GenerateNodeName("WaitEvent"),
+                  "WaitEvent",
+                  "",
+                  input_args,
+                  output_args,
+                  nullptr,
+                  kMSDomain);
+  } else {
+    // the first stage doesn't have recv_fw. Add Wait for all inputs.
```

Comment:
This applies even if there is no `Recv`. Do we really need to separate `recv_fw==NULL` and `recv_fw!=NULL`?

Reply:
yeah, depending on whether recv_fw is nullptr or not, the input and output of Wait will be different, thus needs to be handled differently.

---
In reply to: [409781157](https://github.com/microsoft/onnxruntime/pull/3513#discussion_r409781157) [](ancestors = 409781157)

Reply:
I don't understand. Can't we treat `Recv` as a normal operator and connect `Recv` to outputs of `WaitEvent`?

Reply:
in thise case, there is no Recv op in forward pass (first stage), so instead of conect Recv to otput of Wait, we inset Wait between input tensor and its consuming ops.

---
In reply to: [409822235](https://github.com/microsoft/onnxruntime/pull/3513#discussion_r409822235) [](ancestors = 409822235)